libnetwork/osl: Namespace.DeleteNeighbor: assorted cleanups, and ignore "non-exist" warnings

thaJeztah opened the pull request:
- relates to / addresses https://github.com/moby/moby/issues/46241

### libnetwork/osl: Namespace.DeleteNeighbor: don't warn on non-existing neighbor

The code ignores these errors, but will unconditionally print a warning;

> If the kernel deletion fails for the neighbor entry still remote it
> from the namespace cache. Otherwise if the neighbor moves back to the
> same host again, kernel update can fail.

Let's reduce noise if the neighbor wasn't found, to prevent logs like:

    Aug 16 13:26:35 master1.local dockerd[4019880]: time="2023-08-16T13:26:35.186662370+02:00" level=warning msg="error while deleting neighbor entry" error="no such file or directory"
    Aug 16 13:26:35 master1.local dockerd[4019880]: time="2023-08-16T13:26:35.366585939+02:00" level=warning msg="error while deleting neighbor entry" error="no such file or directory"
    Aug 16 13:26:42 master1.local dockerd[4019880]: time="2023-08-16T13:26:42.366658513+02:00" level=warning msg="error while deleting neighbor entry" error="no such file or directory"

### libnetwork/osl: Namespace.DeleteNeighbor: remove osDelete argument

This argument was originally added in libnetwork:
https://github.com/moby/libnetwork/commit/03f440667f79f33ef3cb1faeb13c0bca1a0f8737

At the time, this argument was conditional, but currently it's always set
to "true", so let's remove it.

### libnetwork/osl: Namespace.DeleteNeighbor: remove intermediate vars

- store linkIndex in a local variable so that it can be reused
- remove / rename some intermediate vars that shadowed existing declaration





**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah committed:
libnetwork/osl: Namespace.DeleteNeighbor: don't warn on non-existing neighbor

The code ignores these errors, but will unconditionally print a warning;

> If the kernel deletion fails for the neighbor entry still remote it
> from the namespace cache. Otherwise if the neighbor moves back to the
> same host again, kernel update can fail.

Let's reduce noise if the neighbor wasn't found, to prevent logs like:

    Aug 16 13:26:35 master1.local dockerd[4019880]: time="2023-08-16T13:26:35.186662370+02:00" level=warning msg="error while deleting neighbor entry" error="no such file or directory"
    Aug 16 13:26:35 master1.local dockerd[4019880]: time="2023-08-16T13:26:35.366585939+02:00" level=warning msg="error while deleting neighbor entry" error="no such file or directory"
    Aug 16 13:26:42 master1.local dockerd[4019880]: time="2023-08-16T13:26:42.366658513+02:00" level=warning msg="error while deleting neighbor entry" error="no such file or directory"

While changing this code, also slightly rephrase the code-comment, and
fix a typo ("remote -> remove").

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

libnetwork/osl: Namespace.DeleteNeighbor: rephrase code-comment

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork/osl: Namespace.DeleteNeighbor: remove osDelete argument

This argument was originally added in libnetwork:
https://github.com/moby/libnetwork/commit/03f440667f79f33ef3cb1faeb13c0bca1a0f8737

At the time, this argument was conditional, but currently it's always set
to "true", so let's remove it.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
libnetwork/osl: Namespace.DeleteNeighbor: remove intermediate vars

- store linkIndex in a local variable so that it can be reused
- remove / rename some intermediate vars that shadowed existing declaration

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

akerouanton commented on libnetwork/osl/neigh_linux.go:
```diff
@@ -43,79 +45,66 @@ func (n *Namespace) findNeighbor(dstIP net.IP, dstMac net.HardwareAddr) *neigh {
+	// If the kernel deletion fails for the neighbor entry still remote it
+	// from the namespace cache. Otherwise if the neighbor moves back to the
+	// same host again, kernel update can fail.
+	if err := nlh.NeighDel(nlnh); err != nil && !errors.Is(err, os.ErrNotExist) {
```
Mh I would tend to say: if it doesn't exist anymore, the end state has been reached anyway. But... this code is called by the `peerdb` package, and:

1. I'm wondering if that could be a symptom of some broken internal state;
2. [this comment](https://github.com/moby/moby/blob/2243046bf8e9b677309744863af116932f9aeabe/libnetwork/drivers/overlay/peerdb.go#L232-L242) references some "message", channel and goroutin but I don't see anything like that hanging around in that method, in its caller or in `peerInitOp`. I guess this comment is outdated, which makes me wonder what prevent two "peer delete" / "peer add" to be process concurrently :thinking: I admit the last time I looked into `peerdb` was months ago and I don't recall how it works exactly :woozy_face: 

> // Note also that this method sends a single message on the channel and the go routine on the
> // other side, will atomically loop on the whole table of peers and will program their state
> // in one single atomic operation. This is fundamental to guarantee consistency, and avoid that
> // new peerAdd or peerDelete gets reordered during the sandbox init.

thaJeztah commented:
@akerouanton I addressed your comments, but if you think it'd be better to move the "ignore not found errors" separate, I can move that part to a separate PR

thaJeztah commented:
@akerouanton this one good to go? Also see my comment above

akerouanton reviewed (approved):
LGTM
